### ISSUE TITLE
Ribbon Retry with RestTemplate?

### ISSUE THREAD

**Issue by tkvangorder** [spring-cloud/spring-cloud-netflix#648]
I have been migrating an existing application over to Spring Cloud's service discovery, Ribbon load balancing, and circuit breakers. The application already makes extensive use of the RestTemplate and I have been able to successfully use the load balanced version of the template. However, I have been testing the situation where there are two instances of a service and I drop one of those instances out of operation. I would like the RestTemplate to failover to the next server. From the research I have done, it appears that the fail-over logic exists in the Feign client and when using Zuul. It appears that the LoadBalancedRest template does not have logic for fail-over. In diving into the code, it looks like the RibbonClientHttpRequestFactory is using the netflix RestClient (which appears to have logic for doing retries).

So where do I go from here to get this working? 

I would prefer to not use the Feign client because I would have to sweep A LOT of code.
I had found this [link](http://stackoverflow.com/questions/27468697/spring-cloud-with-resttemplate-ribbon-eureka-retry-when-server-not-available) that suggested using the @Retryable annotation along with @HystrixCommand but this seems like something that should be a part of the load balanced rest template. 

I did some digging into the code for RibbonClientHttpRequestFactory: 

```
		protected ClientHttpResponse executeInternal(HttpHeaders headers)
				throws IOException {
			try {
				addHeaders(headers);
				if (outputStream != null) {
					outputStream.close();
					builder.entity(outputStream.toByteArray());
				}
				HttpRequest request = builder.build();
				HttpResponse response = client.execute(request, config);
				return new RibbonHttpResponse(response);
			} catch (Exception e) {
				throw new IOException(e);
			}

```
It appears that if I override this method and change it to use "client.executeWithLoadBalancer()" that I might be able to leverage the retry logic that is built into the RestClient? This would require me to create my own version of the RibbonClientHttpRequestFactory, I am hoping I am not wandering too far down the rabbit hole. 

Just looking for guidance on the best approach. 

Thanks

**Comment by tkvangorder** [comment]
@spencergibb I was able to get fail-over working in a modified version of RibbonClientHttpRequestFactory. Was there a reason in choosing "client.execute() instead of client.executeWithLoadBalancer? Can we add this capability into the core? I can generate a pull request if you would like.

I had moved this question to stack overflow and I have documented my changes here: 

[Stack OverFlow Link](http://stackoverflow.com/questions/33765049/spring-cloud-getting-retry-working-in-resttemplate/33786574#33786574)
